File: java/money/ch10/src/main/java/com/thesoftwaregorilla/tdd/money/Money.java
package com.thesoftwaregorilla.tdd.money;

public abstract class Money {

    public static Money dollar(int amount) {
        return new Dollar(amount, "USD");
    }

    public static Money franc(int amount) {
        return new Franc(amount, "CHF");
    }


    private final int amount;
    private final String currency;

    protected Money(int amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public abstract Money times(int multiplier);

    @Override
    public boolean equals(Object object) {
        Money money = (Money) object;
        return getClass().equals(object.getClass()) && amount == money.amount;
    }

    // See my note in the DollarTest class. I added a getter for amount because I had a constructor test.
    public int getAmount() {
        return amount;
    }

    public String currency() {
        return currency;
    }

}

File: cs/ch09/Money/Money.cs
namespace TheSoftwareGorilla.TDD.Money;

public abstract class Money
{
    protected int amount;

    public string Currency { get; }

    protected Money(int amount, string currency)
    {
        this.amount = amount;
        Currency = currency;
    }

    public static Money Dollar(int amount)
    {
        return new Dollar(amount, "USD");
    }

    public static Money Franc(int amount)
    {
        return new Franc(amount, "CHF");
    }

    public abstract Money Times(int multiplier);

    public override bool Equals(object? obj)
    {
        return obj is Money money && GetType().Equals(obj.GetType()) && amount == money.amount;
    }


}